tests/fs/api.private.test.ts

```diff
@@ -0,0 +1,284 @@
+    const fs = await emptyFilesystem()
+
+    await fc.assert(
+      fc.asyncProperty(
+        pathSegment(), async data => {
+          const dirpath = path.directory('private', data[0])
+
+          await fs.mkdir(dirpath)
+
+          expect(await fs.exists(dirpath)).toEqual(true)
+        })
+    )
+  })
+
+  it('removes directories it makes', async () => {
+    const fs = await emptyFilesystem()
+
+    await fc.assert(
+      fc.asyncProperty(
+        pathSegment(), async data => {
+          const dirpath = path.directory('private', data[0])
+
+          await fs.mkdir(dirpath)
+          await fs.rm(dirpath)
+
+          expect(await fs.exists(dirpath)).toEqual(false)
+        })
+    )
+  })
+
+  it('writes files to a directory', async () => {
+    const fs = await emptyFilesystem()
+    const dirpath = path.directory('private', 'testDir')
+    await fs.mkdir(dirpath)
+
+    await fc.assert(
+      fc.asyncProperty(
+        fc.tuple(pathSegment(), fileContent()), async data => {
+          const filepath = path.file('private', 'testDir', data[0])
+          const [val] = data[1]
+
+          await fs.write(filepath, val)
+
+          expect(await fs.exists(filepath)).toEqual(true)
+        })
+    )
+  })
+
+  it('lists files written to a directory', async () => {
+    const fs = await emptyFilesystem()
+    const dirpath = path.directory('private', 'testDir')
+    await fs.mkdir(dirpath)
+
+    await fc.assert(
+      fc.asyncProperty(
+        fc.tuple(pathSegment(), fileContent()), async data => {
+          const filepath = path.file('private', 'testDir', data[0])
+          const [val] = data[1]
+
+          await fs.write(filepath, val)
+          const listing = await fs.ls(dirpath)
+
+          expect(data[0] in listing).toEqual(true)
+        })
+    )
+  })
+
+  it('moves files into a directory', async () => {
+    const fs = await emptyFilesystem()
+    const dirpath = path.directory('private', 'testDir')
+    await fs.mkdir(dirpath)
+
+    await fc.assert(
+      fc.asyncProperty(
+        fc.tuple(pathSegmentPair(), fileContent()), async data => {
+          const [from, to] = data[0]
+          const [val] = data[1]
+          const fromPath = path.file('private', from)
+          const toPath = path.file('private', 'testDir', to)
+
+          await fs.write(fromPath, val)
+          await fs.mv(fromPath, toPath)
+          const fromExists = await fs.exists(fromPath)
+          const toExists = await fs.exists(toPath)
+
+          expect(toExists && !fromExists).toEqual(true)
+        })
+    )
+  })
+})
+
+
+/* Filesystem */
+
+const emptyFilesystem = async () => {
+  rootKey = await crypto.aes.genKeyStr()
+  return FileSystem.empty({
+    localOnly: true,
+    permissions: {
+      fs: {
+        public: [path.root()],
+        private: [path.root()]
+      }
+    },
+    rootKey
+  })
+}
+
+/* Paths */
+
+const pathSegment = () => {
+  return fc.hexaString({ minLength: 1, maxLength: 20 })
```

Comment:
This is good, but I'd like it if we had a specified set of "what characters are good" for paths.
That needs some discussion outside of this PR probably. I.e. do we want to allow basically all utf-8 strings, except containing `/`? Do we want to interoperate with existing filesystems, e.g. look valid path strings are for windows/unix/mac? (Probably, not, given `AUX.txt` is not a valid file in windows).

Just noting this here. (Also, what would happen if we were generating arbitrary utf-8 strings here that don't contain slashes? So something like `fc.fullUnicodeString({ minLength: 1, maxLength: 20 }).filter(s => !s.includes("/"))`.)

Reply:
Yeah, that's a great point. One other consideration here is should there be some bias towards a set of characters? Hex strings or ASCII don't seem like the right bias, but some limitations might be good because the tests are slow and we don't want to use up our runs with characters that aren't likely to be used.

Reply:
I'm not 100% sure whether matching the frequency of use with the frequency of generation is a good idea. I'd rather try to choose the tested values from a space that's most likely to uncover bugs, right? If we try to match the frequency of use with the frequency of generation, it'll (I think) result in missing the less frequently used edge-cases.

To resolve this issue, we'd need to converge on what we think are the actually allowed values in file paths, though. So again, I don't think we'll move on this issue, yet. Maybe an internal talk post in the forum might be the best next step. :thinking:

Reply:
Yep, talk post for sure. There's a tradeoff between catching edge cases and being sure things work for the common cases. Biasing towards common cases could definitely mean we miss edge cases, but with the number of runs we can do, we might miss them anyways.

Maybe this is another thing that we could  do is configure per environment. Like sanity checks in the local environment and more expansive tests in a GH action.

Anyways, yes, talk post. :bookmark: :bookmark: